### PR TITLE
All identity features working

### PR DESCRIPTION
**Pull request by efabens**
- Still needs more comments and tests for all the realm work
- Maven android publishing not working either and will need to be reworked
- May need additional refactoring to support earlier android versions as well

### CODE REVIEW

**Review by willmichael — APPROVED** [review]
Looks good overall, just some comments that hopefully help us bring the sdks closer together

File: e3db/src/main/java/com/tozny/e3db/Realm.kt [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+            noteId,
+            Base64.encodeURL(crypto.getPublicKey(privateKey)),
+            Base64.encodeURL(publicSigningKey)
+        )
+        val completeBrokerLogin = brokerClient.completeBrokerLogin(completeBrokerLoginRequest).execute()
+        when {
+          completeBrokerLogin.isSuccessful -> {
+            completeBrokerLogin.body()!!.transferId.let { transferId ->
+              val anonymousNoteCient = getAnonymousNoteClient(privateSigningKey, publicSigningKey, this.apiURL, null, this.certificatePinner)
+              readNote(transferId, null, anonymousNoteCient, privateKey, publicSigningKey) { readNoteResponse ->
```
[the remark this answers is not in the document]
Comment (efabens) [reply e3db/src/main/java/com/tozny/e3db/Realm.kt]: good idea will make that change

File: e3db/src/main/java/com/tozny/e3db/Realm.kt [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+          handler.post(runnable)
+        }
+      }
+    } else { // Post results to current thread (whatever that is)
+      Executor { runnable -> runnable.run() }
+    }
+
+    internal val mapper = ObjectMapper().registerModule(KotlinModule()).configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true)
+
+    fun deriveNoteCreds(realmName: String, userName: String, password: String, noteType: CredentialType): DerivedNoteCreds {
```
[the remark this answers is not in the document]
Comment (efabens) [reply e3db/src/main/java/com/tozny/e3db/Realm.kt]: nope a testing class i was using needed it but it can be reduced in scope

**Review by lkwdwrd — APPROVED** [review]
Some comments as we discussed while reviewing Friday, but in general looks solid. Nice work. Clean up comments and you're all set.

File: publish/android/build.gradle [hunk 3]
```diff
@@ -86,18 +86,59 @@ signing {
...
-      from components.android
-
-      groupId project(":publish").ext.groupId
-      artifactId 'e3db-client-android'
-      version project(":publish").ext.version
-
-    }
-  }
-}
+//publishing {
```
[the remark this answers is not in the document]
Comment (efabens) [reply publish/android/build.gradle:89]: This is going to get refactored in fairly short order so am going to keep for reference

**Pull request merged**